torch/distributed/_tensor/device_mesh.py

```diff
@@ -100,43 +100,88 @@ class DeviceMesh:
 
     device_type: str
     mesh: torch.Tensor
-    _backend: str
 
     def __init__(
         self,
         device_type: str,
         mesh: MeshExprT,
-        dim_groups: Optional[List[ProcessGroup]] = None,
+        init_process_groups: bool = True,
     ) -> None:
         self.device_type = device_type
         self.mesh = (
             mesh.detach()
             if isinstance(mesh, torch.Tensor)
             else torch.tensor(mesh, dtype=torch.int)
         )
+        # always try to create default (world) pg, even if it is not initialized
+        # already. The world pg is used for device mesh identity (rank) on each
+        # process (we need to know if the current global rank is in the mesh or not)
         default_pg = self._get_or_create_default_group()
+        if init_process_groups:
+            self._dim_groups = self._init_process_groups(default_pg)
+
+    def _get_or_create_default_group(self):
+        if not is_initialized():
+            unique_mesh_values = self.mesh.unique(sorted=True)
+            # assume we are using env:// and with env vars set, TODO: add some check here
+            world_size = int(os.getenv("WORLD_SIZE", unique_mesh_values[-1] + 1))
+
+            if unique_mesh_values.numel() != self.mesh.numel():
+                raise RuntimeError(
+                    f"DeviceMesh cannot have duplicate values, but found {self.mesh.tolist()}"
+                )
+
+            # ranks in mesh must start from 0
+            if unique_mesh_values[0] != 0:
+                raise RuntimeError(
+                    "DeviceMesh ranks must start from 0, "
+                    f"but found min rank = {unique_mesh_values[0]}"
+                )
+
+            # mesh must be contiguous (i.e. from 0 to N-1)
+            if 2 * unique_mesh_values.sum().item() != world_size * (world_size - 1):
+                raise RuntimeError(
+                    f"DeviceMesh should have all ranks of WORLD, but found {self.mesh.tolist()}"
+                )
+
+
+            _backend = "gloo" if self.device_type == "cpu" else "nccl"
+            init_process_group(backend=_backend)
+        else:
+            world_size = get_world_size()
+            if self.mesh.numel() > world_size:
+                raise RuntimeError(
+                    f"Mesh should not be bigger than default world size, but found {self.mesh.numel()} ranks!"
+                )
+
+        # TODO: we should do allgather the mesh tensor to ensure every rank have the same mesh value
+
+        # calculate the coordinates of the current global rank on the mesh
+        rank_coords = (self.mesh == get_rank()).nonzero()
```

Comment:
For my learning, for 1D device mesh each rank is coordinate right?

Reply:
Assume the 1D device mesh (it can be a submesh) [1,3] on 4 ranks.
Because `nonzero()` returns a 2-D Tensor, so rank 1 has `rank_coords` be `[[0]]` and rank 3 has `[[1]]`.

Reply:
@fduwjj I think we should revisit how 1d device mesh is constructed from 2d mesh, the current way of doing it is somewhat weird. i.e. for 2-d mesh [[0, 1], [2, 3]], we should generate two 1-D mesh, instead of a single 1-D mesh.

Reply:
Since now we support sub-mesh. So essentially, we split one 2D mesh into two 1D sub-mesh. And we can decide which one to use based on the current rank. Is this something you are saying?

Reply:
Yes, so basically TP really only run in two sub meshes in a SPMD manner, i.e. in two hosts separately, it does not communicate across ranks(that's the parent mesh sync across the hosts).

Reply:
Do you think it makes sense to have a util func inside `device_mesh.py`  to split one 2D or N-D Mesh into lower dim submesh. Or we already have one?

Reply:
we don't have that yet. I think we can add one to TP first, we can think of whether to make this an API for device_mesh.py once we found there're more use cases :)

We should carefully maintain the device_mesh API going forward as it might be used internally too